File: app/src/main/java/com/andyluu/debrief/ui/Redactions.kt
package com.andyluu.debrief.ui

import com.andyluu.debrief.data.RedactionEntity
import com.andyluu.debrief.data.TranscriptWordEntity

internal const val REDACTION_AUDIO_PAD_MS = 150L
internal const val REDACTION_LABEL = "[redacted]"

internal data class RedactionSelection(
    val segmentId: Long,
    val startMs: Long,
    val endMs: Long,
    val text: String,
    val hasExistingRedactions: Boolean = false,
)

internal data class RedactionRange(
    val startMs: Long,
    val endMs: Long,
    val text: String,
)

internal data class RedactedWordChoice(
    val index: Int,
    val startMs: Long,
    val endMs: Long,
    val text: String,
)

internal fun redactionActiveAt(
    positionMs: Long,
    redactions: List<RedactionEntity>,
    padMs: Long = REDACTION_AUDIO_PAD_MS,
): Boolean = redactions.any { positionMs in (it.startMs - padMs)..(it.endMs + padMs) }

internal fun List<RedactionEntity>.overlappingRedactions(startMs: Long, endMs: Long): List<RedactionEntity> =
    filter { it.startMs < endMs && it.endMs > startMs }

internal fun wordsForSegment(
    words: List<TranscriptWordEntity>,
    segmentStartMs: Long,
    segmentEndMs: Long,
): List<TranscriptWordEntity> =
    words.filter { it.endMs >= segmentStartMs && it.startMs <= segmentEndMs }.sortedBy { it.startMs }

internal fun redactedTranscriptText(
    text: String,
    words: List<TranscriptWordEntity>,
    redactions: List<RedactionEntity>,
    segmentStartMs: Long,
    segmentEndMs: Long,
): String {
    val overlapping = redactions.overlappingRedactions(segmentStartMs, segmentEndMs)
    if (overlapping.isEmpty()) return text
    if (overlapping.any { it.startMs <= segmentStartMs && it.endMs >= segmentEndMs }) return REDACTION_LABEL
    val ranges = wordTextRanges(text, words)
    if (ranges.isEmpty()) return REDACTION_LABEL
    val redactedRanges = ranges.filter { range -> overlapping.any { it.startMs < range.endMs && it.endMs > range.startMs } }
    if (redactedRanges.isEmpty()) return text
    if (redactedRanges.size == ranges.size) return REDACTION_LABEL

    val builder = StringBuilder()
    var cursor = 0
    var lastWasRedacted = false
    ranges.forEach { range ->
        val shouldRedact = redactedRanges.any { it.textStart == range.textStart && it.textEnd == range.textEnd }
        if (shouldRedact) {
            if (!lastWasRedacted) {
                builder.append(text.substring(cursor, range.textStart).trimEnd())
                if (builder.isNotEmpty() && !builder.last().isWhitespace()) builder.append(' ')
                builder.append(REDACTION_LABEL)
            }
            cursor = range.textEnd
            lastWasRedacted = true
        } else {
            val between = text.substring(cursor, range.textStart)
            builder.append(if (lastWasRedacted) between.afterRedaction() else between)
            builder.append(text.substring(range.textStart, range.textEnd))
            cursor = range.textEnd
            lastWasRedacted = false
        }
    }
    val suffix = text.substring(cursor)
    builder.append(if (lastWasRedacted) suffix.afterRedaction() else suffix)
    return builder.toString().replace(Regex("""\s{2,}"""), " ").trim()
}

internal fun segmentRedactionSelection(
    segmentId: Long,
    startMs: Long,
    endMs: Long,
    text: String,
    hasExistingRedactions: Boolean = false,
): RedactionSelection =
    RedactionSelection(segmentId, startMs, endMs, text, hasExistingRedactions)

internal fun redactionRangesForWholeSegment(
    words: List<TranscriptWordEntity>,
    segmentStartMs: Long,
    segmentEndMs: Long,
    fallbackText: String,
): List<RedactionRange> {
    val segmentWords = wordsForSegment(words, segmentStartMs, segmentEndMs)
    if (segmentWords.isEmpty()) return listOf(RedactionRange(segmentStartMs, segmentEndMs, fallbackText))
    return segmentWords.map { word -> RedactionRange(word.startMs, word.endMs, word.text) }
}

internal fun redactedWordChoices(
    words: List<TranscriptWordEntity>,
    redactions: List<RedactionEntity>,
    segmentStartMs: Long,
    segmentEndMs: Long,
): List<RedactedWordChoice> {
    val overlapping = redactions.overlappingRedactions(segmentStartMs, segmentEndMs)
    if (overlapping.isEmpty()) return emptyList()
    val segmentWords = wordsForSegment(words, segmentStartMs, segmentEndMs)
    if (segmentWords.isEmpty()) return emptyList()
    val wholeSegmentRedacted = overlapping.any { it.startMs <= segmentStartMs && it.endMs >= segmentEndMs }
    return segmentWords.mapIndexedNotNull { index, word ->
        val wordRedacted = wholeSegmentRedacted || overlapping.any { it.startMs < word.endMs && it.endMs > word.startMs }
        if (wordRedacted) RedactedWordChoice(index + 1, word.startMs, word.endMs, word.text) else null
    }
}

internal fun redactionRangesAfterRemovingWord(
    words: List<TranscriptWordEntity>,
    redactions: List<RedactionEntity>,
    segmentStartMs: Long,
    segmentEndMs: Long,
    removedWord: RedactedWordChoice,
): List<RedactionRange> {
    val overlapping = redactions.overlappingRedactions(segmentStartMs, segmentEndMs)
    if (overlapping.isEmpty()) return emptyList()
    val segmentWords = wordsForSegment(words, segmentStartMs, segmentEndMs)
    if (segmentWords.isEmpty()) return emptyList()
    val wholeSegmentRedacted = overlapping.any { it.startMs <= segmentStartMs && it.endMs >= segmentEndMs }
    return segmentWords
        .filterNot { it.startMs == removedWord.startMs && it.endMs == removedWord.endMs }
        .filter { word -> wholeSegmentRedacted || overlapping.any { it.startMs < word.endMs && it.endMs > word.startMs } }
        .map { word -> RedactionRange(word.startMs, word.endMs, word.text) }
}

private data class TimedTextRange(
    val textStart: Int,
    val textEnd: Int,
    val startMs: Long,
    val endMs: Long,
    val text: String,
)

private fun wordTextRanges(text: String, words: List<TranscriptWordEntity>): List<TimedTextRange> {
    if (text.isBlank() || words.isEmpty()) return emptyList()
    val tokenMatches = Regex("""\S+""").findAll(text).toList()
    if (tokenMatches.isEmpty()) return emptyList()
    return tokenMatches.zip(words).map { (match, word) ->
        TimedTextRange(
            textStart = match.range.first,
            textEnd = match.range.last + 1,
            startMs = word.startMs,
            endMs = word.endMs,
            text = match.value,
        )
    }
}

private fun String.prependIfNeeded(): String =
    if (isNotEmpty() && first().isWhitespace()) this else if (isNotEmpty()) " $this" else this

private fun String.afterRedaction(): String {
    val stripped = dropWhile(Char::isWhitespace)
    return if (isNotEmpty() && first().isWhitespace()) " $stripped" else stripped.prependIfNeeded()
}

File: web/src/core/redactions.ts
import type { Redaction, TranscriptWord } from './models'

/**
 * Ported from `app/src/main/java/com/andyluu/debrief/ui/Redactions.kt`.
 *
 * Redaction is reversible and never touches the source audio. Text is masked as
 * `[redacted]` and playback is muted across the redacted timestamps, so the
 * screen and the speakers can both be shared safely.
 *
 * The asymmetric padding is deliberate and load-bearing. Provider word starts
 * can land slightly after the audible consonant, and at 4x playback the player
 * advances about 300 ms during one 75 ms volume poll. Favour privacy by muting
 * well before the stored start while keeping a smaller tail.
 */
export const REDACTION_AUDIO_LEAD_PAD_MS = 750
export const REDACTION_AUDIO_TRAIL_PAD_MS = 250
export const REDACTION_LABEL = '[redacted]'

export interface RedactionRange {
  startMs: number
  endMs: number
  text: string
}

export interface RedactedWordChoice {
  index: number
  startMs: number
  endMs: number
  text: string
}

export interface RedactionMuteRange {
  startMs: number
  endMs: number
}

/**
 * Pads every redaction, then merges anything that now overlaps, so the player
 * only has to test a small sorted list on each poll.
 */
export function redactionMuteRanges(
  redactions: Redaction[],
  leadPadMs: number = REDACTION_AUDIO_LEAD_PAD_MS,
  trailPadMs: number = REDACTION_AUDIO_TRAIL_PAD_MS,
): RedactionMuteRange[] {
  const padded = redactions
    .map((redaction) => {
      const start = Math.min(redaction.startMs, redaction.endMs)
      const end = Math.max(redaction.startMs, redaction.endMs)
      return {
        startMs: Math.max(0, start - Math.max(0, leadPadMs)),
        endMs: end + Math.max(0, trailPadMs),
      }
    })
    .sort((a, b) => a.startMs - b.startMs)

  if (padded.length === 0) return []

  const merged: RedactionMuteRange[] = [padded[0]!]
  for (const next of padded.slice(1)) {
    const current = merged[merged.length - 1]!
    if (next.startMs <= current.endMs) {
      merged[merged.length - 1] = { ...current, endMs: Math.max(current.endMs, next.endMs) }
    } else {
      merged.push(next)
    }
  }
  return merged
}

export function redactionActiveAt(
  positionMs: number,
  redactions: Redaction[],
  leadPadMs: number = REDACTION_AUDIO_LEAD_PAD_MS,
  trailPadMs: number = REDACTION_AUDIO_TRAIL_PAD_MS,
): boolean {
  return redactionMuteActiveAt(positionMs, redactionMuteRanges(redactions, leadPadMs, trailPadMs))
}

export function redactionPlaybackVolumeAt(positionMs: number, redactions: Redaction[]): number {
  return redactionPlaybackVolumeForRanges(positionMs, redactionMuteRanges(redactions))
}

export function redactionMuteActiveAt(
  positionMs: number,
  muteRanges: RedactionMuteRange[],
): boolean {
  return muteRanges.some((range) => positionMs >= range.startMs && positionMs <= range.endMs)
}

export function redactionPlaybackVolumeForRanges(
  positionMs: number,
  muteRanges: RedactionMuteRange[],
): number {
  return redactionMuteActiveAt(positionMs, muteRanges) ? 0 : 1
}

export function overlappingRedactions(
  redactions: Redaction[],
  startMs: number,
  endMs: number,
): Redaction[] {
  return redactions.filter((r) => r.startMs < endMs && r.endMs > startMs)
}

export function wordsForSegment(
  words: TranscriptWord[],
  segmentStartMs: number,
  segmentEndMs: number,
): TranscriptWord[] {
  return words
    .filter((word) => word.endMs >= segmentStartMs && word.startMs <= segmentEndMs)
    .sort((a, b) => a.startMs - b.startMs)
}

interface TimedTextRange {
  textStart: number
  textEnd: number
  startMs: number
  endMs: number
  text: string
}

/**
 * Masks the redacted words inside one transcript card.
 *
 * Consecutive redacted words collapse into a single `[redacted]` so the card
 * does not leak how many words were hidden. When every word is redacted - or
 * when word timing is missing entirely - the whole card becomes one label.
 */
export function redactedTranscriptText(
  text: string,
  words: TranscriptWord[],
  redactions: Redaction[],
  segmentStartMs: number,
  segmentEndMs: number,
): string {
  const overlapping = overlappingRedactions(redactions, segmentStartMs, segmentEndMs)
  if (overlapping.length === 0) return text
  if (overlapping.some((r) => r.startMs <= segmentStartMs && r.endMs >= segmentEndMs)) {
    return REDACTION_LABEL
  }

  const ranges = wordTextRanges(text, words)
  // No word timing means we cannot mask precisely, so mask everything rather
  // than risk showing something that should have been hidden.
  if (ranges.length === 0) return REDACTION_LABEL

  const redactedRanges = ranges.filter((range) =>
    overlapping.some((r) => r.startMs < range.endMs && r.endMs > range.startMs),
  )
  if (redactedRanges.length === 0) return text
  if (redactedRanges.length === ranges.length) return REDACTION_LABEL

  let builder = ''
  let cursor = 0
  let lastWasRedacted = false

  for (const range of ranges) {
    const shouldRedact = redactedRanges.some(
      (r) => r.textStart === range.textStart && r.textEnd === range.textEnd,
    )
    if (shouldRedact) {
      if (!lastWasRedacted) {
        builder += trimEnd(text.slice(cursor, range.textStart))
        if (builder.length > 0 && !/\s/.test(builder[builder.length - 1]!)) builder += ' '
        builder += REDACTION_LABEL
      }
      cursor = range.textEnd
      lastWasRedacted = true
    } else {
      const between = text.slice(cursor, range.textStart)
      builder += lastWasRedacted ? afterRedaction(between) : between
      builder += text.slice(range.textStart, range.textEnd)
      cursor = range.textEnd
      lastWasRedacted = false
    }
  }

  const suffix = text.slice(cursor)
  builder += lastWasRedacted ? afterRedaction(suffix) : suffix
  return builder.replace(/\s{2,}/g, ' ').trim()
}

/** Every word of a card, for the "redact this whole card" action. */
export function redactionRangesForWholeSegment(
  words: TranscriptWord[],
  segmentStartMs: number,
  segmentEndMs: number,
  fallbackText: string,
): RedactionRange[] {
  const segmentWords = wordsForSegment(words, segmentStartMs, segmentEndMs)
  if (segmentWords.length === 0) {
    return [{ startMs: segmentStartMs, endMs: segmentEndMs, text: fallbackText }]
  }
  return segmentWords.map((word) => ({ startMs: word.startMs, endMs: word.endMs, text: word.text }))
}

/**
 * The undo menu: which words in this card are currently hidden, numbered from
 * 1, with their real text so the user can pick the one to reveal.
 */
export function redactedWordChoices(
  words: TranscriptWord[],
  redactions: Redaction[],
  segmentStartMs: number,
  segmentEndMs: number,
): RedactedWordChoice[] {
  const overlapping = overlappingRedactions(redactions, segmentStartMs, segmentEndMs)
  if (overlapping.length === 0) return []
  const segmentWords = wordsForSegment(words, segmentStartMs, segmentEndMs)
  if (segmentWords.length === 0) return []

  const wholeSegmentRedacted = overlapping.some(
    (r) => r.startMs <= segmentStartMs && r.endMs >= segmentEndMs,
  )
  const choices: RedactedWordChoice[] = []
  segmentWords.forEach((word, index) => {
    const wordRedacted =
      wholeSegmentRedacted ||
      overlapping.some((r) => r.startMs < word.endMs && r.endMs > word.startMs)
    if (wordRedacted) {
      choices.push({ index: index + 1, startMs: word.startMs, endMs: word.endMs, text: word.text })
    }
  })
  return choices
}

/**
 * Rebuilds a card's redactions with one word revealed. A whole-card redaction
 * is expanded into per-word ranges first, so revealing one word does not
 * un-redact the rest.
 */
export function redactionRangesAfterRemovingWord(
  words: TranscriptWord[],
  redactions: Redaction[],
  segmentStartMs: number,
  segmentEndMs: number,
  removedWord: RedactedWordChoice,
): RedactionRange[] {
  const overlapping = overlappingRedactions(redactions, segmentStartMs, segmentEndMs)
  if (overlapping.length === 0) return []
  const segmentWords = wordsForSegment(words, segmentStartMs, segmentEndMs)
  if (segmentWords.length === 0) return []

  const wholeSegmentRedacted = overlapping.some(
    (r) => r.startMs <= segmentStartMs && r.endMs >= segmentEndMs,
  )
  return segmentWords
    .filter((word) => !(word.startMs === removedWord.startMs && word.endMs === removedWord.endMs))
    .filter(
      (word) =>
        wholeSegmentRedacted ||
        overlapping.some((r) => r.startMs < word.endMs && r.endMs > word.startMs),
    )
    .map((word) => ({ startMs: word.startMs, endMs: word.endMs, text: word.text }))
}

/**
 * Pairs whitespace-delimited tokens in the rendered text with the provider's
 * word list positionally. `zip` semantics: if the two lists disagree in length,
 * the shorter one wins, exactly as Kotlin's `zip` does.
 */
function wordTextRanges(text: string, words: TranscriptWord[]): TimedTextRange[] {
  if (text.trim().length === 0 || words.length === 0) return []
  const tokens = [...text.matchAll(/\S+/g)]
  if (tokens.length === 0) return []

  const pairCount = Math.min(tokens.length, words.length)
  const ranges: TimedTextRange[] = []
  for (let index = 0; index < pairCount; index += 1) {
    const match = tokens[index]!
    const word = words[index]!
    const textStart = match.index
    ranges.push({
      textStart,
      textEnd: textStart + match[0].length,
      startMs: word.startMs,
      endMs: word.endMs,
      text: match[0],
    })
  }
  return ranges
}

function trimEnd(value: string): string {
  return value.replace(/\s+$/, '')
}

/**
 * Rejoins text after a `[redacted]` label, collapsing the whitespace that the
 * removed words left behind into exactly one separating space.
 */
function afterRedaction(value: string): string {
  const stripped = value.replace(/^\s+/, '')
  if (value.length > 0 && /\s/.test(value[0]!)) return ` ${stripped}`
  return prependIfNeeded(stripped)
}

function prependIfNeeded(value: string): string {
  if (value.length > 0 && /\s/.test(value[0]!)) return value
  if (value.length > 0) return ` ${value}`
  return value
}
